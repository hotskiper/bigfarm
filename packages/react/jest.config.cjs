module.exports = {
    roots: ['<rootDir>/src'],
    testRegex: '(/.*\\.test)\\.(ts|tsx)$',
    setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
    moduleNameMapper: {
        '^@/(.*)$': '<rootDir>/src/$1'
    },
    testEnvironment: 'jsdom'
}