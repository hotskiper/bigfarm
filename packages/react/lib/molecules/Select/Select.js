import React, { useState, useRef, useEffect, createRef } from 'react';
import Text from '../../atoms/Text/Text.js';

const getPreviousOptionIndex = (currentIndex, options) => {
    if (currentIndex === null) {
        return 0;
    }
    if (currentIndex === 0) {
        return options.length - 1;
    }
    return currentIndex - 1;
};
const getNextOptionIndex = (currentIndex, options) => {
    if (currentIndex === null) {
        return 0;
    }
    if (currentIndex === options.length - 1) {
        return 0;
    }
    return currentIndex + 1;
};
const Select = ({ options = [], label = "Please select", onOptionSelected: handler, renderOption, }) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selectedIndex, setSelectedIndex] = useState(null);
    const [highlightedIndex, setHighlightedIndex] = useState(null);
    const [overlayTop, setOverlayTop] = useState(0);
    const [optionRefs, setOptionRefs] = useState([]);
    const labelRef = useRef(null);
    useEffect(() => {
        setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
    }, [labelRef.current?.offsetHeight]);
    const onOptionSelected = (option, optionIndex) => {
        if (handler)
            handler(option, optionIndex);
        setSelectedIndex(optionIndex);
        setIsOpen(false);
    };
    const onLabelClick = () => {
        setIsOpen(!isOpen);
    };
    const highlightOption = (optionIndex) => {
        setHighlightedIndex(optionIndex);
    };
    const onButtonKeyDown = (event) => {
        event.preventDefault();
        if (['Enter', '', 'ArrowDown'].includes(event.key)) {
            setIsOpen(true);
            highlightOption(0);
        }
    };
    useEffect(() => {
        setOptionRefs(options.map((_) => createRef()));
    }, [options.length]);
    useEffect(() => {
        if (highlightedIndex !== null && isOpen) {
            const ref = optionRefs[highlightedIndex];
            if (ref && ref.current) {
                ref.current.focus();
            }
        }
    }, [isOpen, highlightedIndex]);
    // console.log(1, optionRefs);
    const onOptionKeyDown = (event) => {
        console.log(event.key);
        if (event.key === 'Escape') {
            setIsOpen(false);
            return;
        }
        if (event.key === 'ArrowDown') {
            highlightOption(getNextOptionIndex(highlightedIndex, options));
        }
        if (event.key === 'ArrowUp') {
            highlightOption(getPreviousOptionIndex(highlightedIndex, options));
        }
        if (event.key === 'Enter') {
            onOptionSelected(options[highlightedIndex], highlightedIndex);
        }
    };
    return (React.createElement("div", { className: 'dse-select' },
        React.createElement("button", { "data-testid": 'DseSelectButton', role: 'menu', "aria-controls": 'dse-select-list', "aria-haspopup": true, "aria-expanded": isOpen ? true : undefined, ref: labelRef, className: 'dse-select__label', onKeyDown: onButtonKeyDown, onClick: () => {
                onLabelClick();
            } },
            React.createElement(Text, null, selectedIndex === null
                ? label
                : options[selectedIndex].label),
            React.createElement("svg", { width: '1rem', height: '1rem', xmlns: 'http://www.w3.org/2000/svg', fill: 'none', viewBox: '0 0 24 24', strokeWidth: 1.5, stroke: 'currentColor', className: `dse-select__caret ${isOpen
                    ? "dse-select__caret--open"
                    : "dse-select__caret--closed"}` },
                React.createElement("path", { strokeLinecap: 'round', strokeLinejoin: 'round', d: 'm4.5 15.75 7.5-7.5 7.5 7.5' }))),
        isOpen && (React.createElement("ul", { id: 'dse-select-list', style: { top: overlayTop }, className: 'dse-select__overlay' }, options.map((option, optionIndex) => {
            const isSelected = optionIndex === selectedIndex;
            const isHighlighted = optionIndex === highlightedIndex;
            const ref = optionRefs[optionIndex];
            const renderOptionProps = {
                isSelected,
                option,
                getOptionRecommendedProps: (overrideProps = {}) => {
                    return {
                        ref,
                        'aria-checked': isSelected ? true : undefined,
                        'aria-label': option.label,
                        tabIndex: isHighlighted ? -1 : 0,
                        role: "menuitemradio",
                        className: `dse-select__option 
                                        ${isSelected ? "dse-select__option--selected" : ""} 
                                        ${isHighlighted ? "dse-select__option--highlighted" : ""}
                                    `,
                        key: option.value,
                        onMouseEnter: () => highlightOption(optionIndex),
                        onMouseLeave: () => highlightOption(null),
                        onKeyDown: onOptionKeyDown,
                        onClick: () => onOptionSelected(option, optionIndex),
                        ...overrideProps,
                    };
                },
            };
            if (renderOption) {
                return renderOption(renderOptionProps);
            }
            return (React.createElement("li", { ...renderOptionProps.getOptionRecommendedProps() },
                React.createElement(Text, null, option.label),
                isSelected && (React.createElement("svg", { width: '1rem', height: '1rem', xmlns: 'http://www.w3.org/2000/svg', fill: 'none', viewBox: '0 0 24 24', strokeWidth: 1.5, stroke: 'currentColor', className: 'w-6 h-6' },
                    React.createElement("path", { strokeLinecap: 'round', strokeLinejoin: 'round', d: 'm4.5 12.75 6 6 9-13.5' })))));
        })))));
};

export { Select as default };
//# sourceMappingURL=Select.js.map
