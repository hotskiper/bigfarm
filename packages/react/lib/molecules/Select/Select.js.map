{"version":3,"file":"Select.js","sources":["../../../src/molecules/Select/Select.tsx"],"sourcesContent":["import React, { useState, useRef, useEffect, KeyboardEventHandler } from \"react\";\r\nimport Text from \"@/atoms/Text\";\r\n\r\nconst KEY_CODES = {\r\n    ENTER: 13,\r\n    SPACE: 32,\r\n    DOWN_ARROW: 40\r\n}\r\n\r\ninterface SelectOption {\r\n    label: string;\r\n    value: string;\r\n}\r\n\r\ninterface RenderOptionProps {\r\n    isSelected: boolean;\r\n    option: SelectOption;\r\n    getOptionRecommendedProps: (overrideProps?: Object) => Object;\r\n}\r\n\r\ninterface SelectProps {\r\n    onOptionSelected?: (option: SelectOption, optionIndex: number) => void;\r\n    options?: SelectOption[];\r\n    label?: string;\r\n    renderOption?: (props: RenderOptionProps) => React.ReactNode;\r\n}\r\n\r\nconst Select: React.FC<SelectProps> = ({\r\n    options = [],\r\n    label = \"Please select\",\r\n    onOptionSelected: handler,\r\n    renderOption,\r\n}) => {\r\n    const [isOpen, setIsOpen] = useState<boolean>(false);\r\n    const [selectedIndex, setSelectedIndex] = useState<null | number>(null);\r\n    const [overlayTop, setOverlayTop] = useState<number>(0);\r\n    const [optionRefs, setOptionRefs] = useState<React.RefObject<HTMLLIElement>[]>([])\r\n    const labelRef = useRef<HTMLButtonElement>(null);\r\n\r\n    useEffect(() => {\r\n        setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);\r\n    }, [labelRef.current?.offsetHeight]);\r\n\r\n    const onOptionSelected = (option: SelectOption, optionIndex: number) => {\r\n        if (handler) handler(option, optionIndex);\r\n        setSelectedIndex(optionIndex);\r\n        setIsOpen(false);\r\n    };\r\n\r\n    const onLabelClick = () => {\r\n        setIsOpen(!isOpen);\r\n    };\r\n \r\n    const onButtonKeyDown: KeyboardEventHandler = (event) => {\r\n        event.preventDefault();\r\n        if([KEY_CODES.ENTER, KEY_CODES.SPACE, KEY_CODES.DOWN_ARROW].includes(event.keyCode)){\r\n            setIsOpen(true)\r\n        }\r\n    }\r\n\r\n    return (\r\n        <div className='dse-select'>\r\n            <button \r\n            role=\"menu\"\r\n            aria-controls=\"dse-select-list\"\r\n                aria-haspopup={true}\r\n                aria-expanded={isOpen?true:undefined}\r\n                ref={labelRef}\r\n                className='dse-select__label'\r\n                onKeyDown={onButtonKeyDown}\r\n                onClick={() => {\r\n                    onLabelClick();\r\n                }}\r\n            >\r\n                <Text>\r\n                    {selectedIndex === null\r\n                        ? label\r\n                        : options[selectedIndex].label}\r\n                </Text>\r\n\r\n                <svg\r\n                    width='1rem'\r\n                    height='1rem'\r\n                    xmlns='http://www.w3.org/2000/svg'\r\n                    fill='none'\r\n                    viewBox='0 0 24 24'\r\n                    strokeWidth={1.5}\r\n                    stroke='currentColor'\r\n                    className={`dse-select__caret ${\r\n                        isOpen\r\n                            ? \"dse-select__caret--open\"\r\n                            : \"dse-select__caret--closed\"\r\n                    }`}\r\n                >\r\n                    <path\r\n                        strokeLinecap='round'\r\n                        strokeLinejoin='round'\r\n                        d='m4.5 15.75 7.5-7.5 7.5 7.5'\r\n                    />\r\n                </svg>\r\n            </button>\r\n            {isOpen && (\r\n                <ul id=\"dse-select-list\" style={{ top: overlayTop }} className='dse-select__overlay'>\r\n                    {options.map((option, optionIndex) => {\r\n                        const isSelected = optionIndex === selectedIndex;\r\n                        const renderOptionProps = {\r\n                            isSelected,\r\n                            option,\r\n                            getOptionRecommendedProps: (overrideProps = {}) => {\r\n                                return {\r\n                                    className: `dse-select__option ${\r\n                                        isSelected\r\n                                            ? \"dse-select__option--selected\"\r\n                                            : \"\"\r\n                                    }`,\r\n                                    key: option.value,\r\n                                    onClick: ()=> onOptionSelected(option, optionIndex),\r\n                                    ...overrideProps\r\n                                };\r\n                            },\r\n                        };\r\n                        if (renderOption) {\r\n                            return renderOption(renderOptionProps);\r\n                        }\r\n                        return (\r\n                            <li\r\n                                className={`dse-select__option ${\r\n                                    isSelected\r\n                                        ? \"dse-select__option--selected\"\r\n                                        : \"\"\r\n                                }`}\r\n                                onClick={() => {\r\n                                    onOptionSelected(option, optionIndex);\r\n                                }}\r\n                                key={option.value}\r\n                            >\r\n                                <Text>{option.label}</Text>\r\n                                {isSelected && (\r\n                                    <svg\r\n                                        width='1rem'\r\n                                        height='1rem'\r\n                                        xmlns='http://www.w3.org/2000/svg'\r\n                                        fill='none'\r\n                                        viewBox='0 0 24 24'\r\n                                        strokeWidth={1.5}\r\n                                        stroke='currentColor'\r\n                                        className='w-6 h-6'\r\n                                    >\r\n                                        <path\r\n                                            strokeLinecap='round'\r\n                                            strokeLinejoin='round'\r\n                                            d='m4.5 12.75 6 6 9-13.5'\r\n                                        />\r\n                                    </svg>\r\n                                )}\r\n                            </li>\r\n                        );\r\n                    })}\r\n                </ul>\r\n            )}\r\n        </div>\r\n    );\r\n};\r\n\r\nexport default Select;\r\n"],"names":[],"mappings":";;;AAGA,MAAM,SAAS,GAAG;AACd,IAAA,KAAK,EAAE,EAAE;AACT,IAAA,KAAK,EAAE,EAAE;AACT,IAAA,UAAU,EAAE,EAAE;CACjB,CAAA;AAoBD,MAAM,MAAM,GAA0B,CAAC,EACnC,OAAO,GAAG,EAAE,EACZ,KAAK,GAAG,eAAe,EACvB,gBAAgB,EAAE,OAAO,EACzB,YAAY,GACf,KAAI;IACD,MAAM,CAAC,MAAM,EAAE,SAAS,CAAC,GAAG,QAAQ,CAAU,KAAK,CAAC,CAAC;IACrD,MAAM,CAAC,aAAa,EAAE,gBAAgB,CAAC,GAAG,QAAQ,CAAgB,IAAI,CAAC,CAAC;IACxE,MAAM,CAAC,UAAU,EAAE,aAAa,CAAC,GAAG,QAAQ,CAAS,CAAC,CAAC,CAAC;IACpB,QAAQ,CAAmC,EAAE,EAAC;AAClF,IAAA,MAAM,QAAQ,GAAG,MAAM,CAAoB,IAAI,CAAC,CAAC;IAEjD,SAAS,CAAC,MAAK;AACX,QAAA,aAAa,CAAC,CAAC,QAAQ,CAAC,OAAO,EAAE,YAAY,IAAI,CAAC,IAAI,EAAE,CAAC,CAAC;KAC7D,EAAE,CAAC,QAAQ,CAAC,OAAO,EAAE,YAAY,CAAC,CAAC,CAAC;AAErC,IAAA,MAAM,gBAAgB,GAAG,CAAC,MAAoB,EAAE,WAAmB,KAAI;AACnE,QAAA,IAAI,OAAO;AAAE,YAAA,OAAO,CAAC,MAAM,EAAE,WAAW,CAAC,CAAC;QAC1C,gBAAgB,CAAC,WAAW,CAAC,CAAC;QAC9B,SAAS,CAAC,KAAK,CAAC,CAAC;AACrB,KAAC,CAAC;IAEF,MAAM,YAAY,GAAG,MAAK;AACtB,QAAA,SAAS,CAAC,CAAC,MAAM,CAAC,CAAC;AACvB,KAAC,CAAC;AAEF,IAAA,MAAM,eAAe,GAAyB,CAAC,KAAK,KAAI;QACpD,KAAK,CAAC,cAAc,EAAE,CAAC;QACvB,IAAG,CAAC,SAAS,CAAC,KAAK,EAAE,SAAS,CAAC,KAAK,EAAE,SAAS,CAAC,UAAU,CAAC,CAAC,QAAQ,CAAC,KAAK,CAAC,OAAO,CAAC,EAAC;YAChF,SAAS,CAAC,IAAI,CAAC,CAAA;SAClB;AACL,KAAC,CAAA;AAED,IAAA,QACI,KAAA,CAAA,aAAA,CAAA,KAAA,EAAA,EAAK,SAAS,EAAC,YAAY,EAAA;AACvB,QAAA,KAAA,CAAA,aAAA,CAAA,QAAA,EAAA,EACA,IAAI,EAAC,MAAM,EAAA,eAAA,EACG,iBAAiB,EACZ,eAAA,EAAA,IAAI,EACJ,eAAA,EAAA,MAAM,GAAC,IAAI,GAAC,SAAS,EACpC,GAAG,EAAE,QAAQ,EACb,SAAS,EAAC,mBAAmB,EAC7B,SAAS,EAAE,eAAe,EAC1B,OAAO,EAAE,MAAK;AACV,gBAAA,YAAY,EAAE,CAAC;aAClB,EAAA;AAED,YAAA,KAAA,CAAA,aAAA,CAAC,IAAI,EAAA,IAAA,EACA,aAAa,KAAK,IAAI;AACnB,kBAAE,KAAK;AACP,kBAAE,OAAO,CAAC,aAAa,CAAC,CAAC,KAAK,CAC/B;AAEP,YAAA,KAAA,CAAA,aAAA,CAAA,KAAA,EAAA,EACI,KAAK,EAAC,MAAM,EACZ,MAAM,EAAC,MAAM,EACb,KAAK,EAAC,4BAA4B,EAClC,IAAI,EAAC,MAAM,EACX,OAAO,EAAC,WAAW,EACnB,WAAW,EAAE,GAAG,EAChB,MAAM,EAAC,cAAc,EACrB,SAAS,EAAE,qBACP,MAAM;AACF,sBAAE,yBAAyB;sBACzB,2BACV,CAAE,CAAA,EAAA;AAEF,gBAAA,KAAA,CAAA,aAAA,CAAA,MAAA,EAAA,EACI,aAAa,EAAC,OAAO,EACrB,cAAc,EAAC,OAAO,EACtB,CAAC,EAAC,4BAA4B,EAAA,CAChC,CACA,CACD;AACR,QAAA,MAAM,KACH,KAAI,CAAA,aAAA,CAAA,IAAA,EAAA,EAAA,EAAE,EAAC,iBAAiB,EAAC,KAAK,EAAE,EAAE,GAAG,EAAE,UAAU,EAAE,EAAE,SAAS,EAAC,qBAAqB,EAC/E,EAAA,OAAO,CAAC,GAAG,CAAC,CAAC,MAAM,EAAE,WAAW,KAAI;AACjC,YAAA,MAAM,UAAU,GAAG,WAAW,KAAK,aAAa,CAAC;AACjD,YAAA,MAAM,iBAAiB,GAAG;gBACtB,UAAU;gBACV,MAAM;AACN,gBAAA,yBAAyB,EAAE,CAAC,aAAa,GAAG,EAAE,KAAI;oBAC9C,OAAO;wBACH,SAAS,EAAE,sBACP,UAAU;AACN,8BAAE,8BAA8B;8BAC9B,EACV,CAAE,CAAA;wBACF,GAAG,EAAE,MAAM,CAAC,KAAK;wBACjB,OAAO,EAAE,MAAK,gBAAgB,CAAC,MAAM,EAAE,WAAW,CAAC;AACnD,wBAAA,GAAG,aAAa;qBACnB,CAAC;iBACL;aACJ,CAAC;YACF,IAAI,YAAY,EAAE;AACd,gBAAA,OAAO,YAAY,CAAC,iBAAiB,CAAC,CAAC;aAC1C;AACD,YAAA,QACI,KAAA,CAAA,aAAA,CAAA,IAAA,EAAA,EACI,SAAS,EAAE,sBACP,UAAU;AACN,sBAAE,8BAA8B;AAChC,sBAAE,EACV,CAAA,CAAE,EACF,OAAO,EAAE,MAAK;AACV,oBAAA,gBAAgB,CAAC,MAAM,EAAE,WAAW,CAAC,CAAC;AAC1C,iBAAC,EACD,GAAG,EAAE,MAAM,CAAC,KAAK,EAAA;AAEjB,gBAAA,KAAA,CAAA,aAAA,CAAC,IAAI,EAAA,IAAA,EAAE,MAAM,CAAC,KAAK,CAAQ;AAC1B,gBAAA,UAAU,KACP,KACI,CAAA,aAAA,CAAA,KAAA,EAAA,EAAA,KAAK,EAAC,MAAM,EACZ,MAAM,EAAC,MAAM,EACb,KAAK,EAAC,4BAA4B,EAClC,IAAI,EAAC,MAAM,EACX,OAAO,EAAC,WAAW,EACnB,WAAW,EAAE,GAAG,EAChB,MAAM,EAAC,cAAc,EACrB,SAAS,EAAC,SAAS,EAAA;AAEnB,oBAAA,KAAA,CAAA,aAAA,CAAA,MAAA,EAAA,EACI,aAAa,EAAC,OAAO,EACrB,cAAc,EAAC,OAAO,EACtB,CAAC,EAAC,uBAAuB,EAC3B,CAAA,CACA,CACT,CACA,EACP;AACN,SAAC,CAAC,CACD,CACR,CACC,EACR;AACN;;;;"}