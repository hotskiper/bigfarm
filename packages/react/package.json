{
  "name": "@bigfarm/react",
  "version": "1.0.0",
  "main": "lib/index.js",
  "type": "module",
  "license": "MIT",
  "devDependencies": {
    "@babel/preset-env": "^7.23.9",
    "@babel/preset-react": "^7.23.3",
    "@babel/preset-typescript": "^7.23.3",
    "@storybook/addon-a11y": "^7.6.17",
    "@storybook/addon-essentials": "^7.6.16",
    "@storybook/addon-interactions": "^7.6.16",
    "@storybook/addon-knobs": "^7.0.2",
    "@storybook/addon-links": "^7.6.16",
    "@storybook/addon-onboarding": "1.0.11",
    "@storybook/addon-storysource": "^7.6.16",
    "@storybook/blocks": "^7.6.16",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^7.6.16",
    "@storybook/react-vite": "^7.6.16",
    "@storybook/test": "^7.6.16",
    "@testing-library/jest-dom": "^6.4.2",
    "@testing-library/react": "^14.2.1",
    "@types/jest": "^29.5.12",
    "@types/react": "^18.2.51",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rollup": "^4.9.6",
    "rollup-plugin-typescript2": "^0.36.0",
    "storybook": "^7.6.16",
    "typescript": "^5.3.3"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn watch --watch",
    "storybook": "storybook dev -p 6006",
    "build-storybook": "build-storybook"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "1e968abec5e267be425a7545ca6f0c1aa689231b"
}
