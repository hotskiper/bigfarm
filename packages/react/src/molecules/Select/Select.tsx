import React, {
    useState,
    useRef,
    useEffect,
    KeyboardEventHandler,
    createRef,
} from "react";
import Text from "@/atoms/Text";


interface SelectOption {
    label: string;
    value: string;
}

interface RenderOptionProps {
    isSelected: boolean;
    option: SelectOption;
    getOptionRecommendedProps: (overrideProps?: Object) => Object;
}

interface SelectProps {
    onOptionSelected?: (option: SelectOption, optionIndex: number) => void;
    options?: SelectOption[];
    label?: string;
    renderOption?: (props: RenderOptionProps) => React.ReactNode;
}

const getPreviousOptionIndex = (currentIndex:number|null, options: Array<SelectOption>)=>{
    if(currentIndex === null){
        return 0;
    }
    if(currentIndex === 0){
        return options.length - 1;
    }
    return currentIndex - 1;
}

const getNextOptionIndex = (currentIndex:number|null, options: Array<SelectOption>)=>{
    if(currentIndex === null){
        return 0;
    }
    if(currentIndex === options.length - 1){
        return 0;
    }
    return currentIndex + 1;
}

const Select: React.FC<SelectProps> = ({
    options = [],
    label = "Please select",
    onOptionSelected: handler,
    renderOption,
}) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    const [selectedIndex, setSelectedIndex] = useState<null | number>(null);
    const [highlightedIndex, setHighlightedIndex] = useState<null | number>(
        null
    );
    const [overlayTop, setOverlayTop] = useState<number>(0);
    const [optionRefs, setOptionRefs] = useState<
        React.RefObject<HTMLLIElement>[]
    >([]);
    const labelRef = useRef<HTMLButtonElement>(null);

    useEffect(() => {
        setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
    }, [labelRef.current?.offsetHeight]);

    const onOptionSelected = (option: SelectOption, optionIndex: number) => {
        if (handler) handler(option, optionIndex);
        setSelectedIndex(optionIndex);
        setIsOpen(false);
    };

    const onLabelClick = () => {
        setIsOpen(!isOpen);
    };

    const highlightOption = (optionIndex: number | null) => {
        setHighlightedIndex(optionIndex);

    };

    const onButtonKeyDown: KeyboardEventHandler = (event) => {
        event.preventDefault();
        if (
            ['Enter','','ArrowDown'].includes(
                event.key
            )
        ) {
            setIsOpen(true); 
            highlightOption(0)
        }
    };

    useEffect(() => {
        setOptionRefs(options.map((_) => createRef<HTMLLIElement>()));
    }, [options.length]);

    useEffect(()=>{
        if(highlightedIndex !== null && isOpen){
            const ref = optionRefs[highlightedIndex]
            if(ref && ref.current){
                ref.current.focus();
            }
        }
    },[isOpen, highlightedIndex])

    // console.log(1, optionRefs);

   

    const onOptionKeyDown: KeyboardEventHandler = (event) =>{
        console.log(event.key)
        if(event.key === 'Escape'){
            setIsOpen(false)
            return;
        }
        if(event.key === 'ArrowDown'){
            highlightOption(getNextOptionIndex(highlightedIndex, options));
        }
        if(event.key === 'ArrowUp'){
            highlightOption(getPreviousOptionIndex(highlightedIndex, options));
        }
        if(event.key === 'Enter'){
            onOptionSelected(options[highlightedIndex!], highlightedIndex!)
        }
    }

    return (
        <div className='dse-select'>
            <button
                data-testid='DseSelectButton'
                role='menu'
                aria-controls='dse-select-list'
                aria-haspopup={true}
                aria-expanded={isOpen ? true : undefined}
                ref={labelRef}
                className='dse-select__label'
                onKeyDown={onButtonKeyDown}
                onClick={() => {
                    onLabelClick();
                }}
            >
                <Text>
                    {selectedIndex === null
                        ? label
                        : options[selectedIndex].label}
                </Text>

                <svg
                    width='1rem'
                    height='1rem'
                    xmlns='http://www.w3.org/2000/svg'
                    fill='none'
                    viewBox='0 0 24 24'
                    strokeWidth={1.5}
                    stroke='currentColor'
                    className={`dse-select__caret ${
                        isOpen
                            ? "dse-select__caret--open"
                            : "dse-select__caret--closed"
                    }`}
                >
                    <path
                        strokeLinecap='round'
                        strokeLinejoin='round'
                        d='m4.5 15.75 7.5-7.5 7.5 7.5'
                    />
                </svg>
            </button>
            {isOpen && (
                <ul
                    id='dse-select-list'
                    style={{ top: overlayTop }}
                    className='dse-select__overlay'
                >
                    {options.map((option, optionIndex) => {
                        const isSelected = optionIndex === selectedIndex;
                        const isHighlighted = optionIndex === highlightedIndex;

                        const ref = optionRefs[optionIndex];

                        const renderOptionProps = {
                            isSelected,
                            option,
                            getOptionRecommendedProps: (overrideProps = {}) => {
                                return {
                                    ref,
                                    'aria-checked': isSelected ? true : undefined,
                                    'aria-label': option.label,
                                    tabIndex: isHighlighted ? -1: 0,
                                    role: "menuitemradio",
                                    className: `dse-select__option 
                                        ${isSelected ? "dse-select__option--selected" : ""} 
                                        ${isHighlighted ? "dse-select__option--highlighted" : ""}
                                    `,
                                    key: option.value,
                                    onMouseEnter: () =>
                                        highlightOption(optionIndex),
                                    onMouseLeave: () => highlightOption(null),
                                    onKeyDown: onOptionKeyDown,
                                    onClick: () =>
                                        onOptionSelected(option, optionIndex),
                                    ...overrideProps,
                                };
                            },
                        };
                        if (renderOption) {
                            return renderOption(renderOptionProps);
                        }
                        return (
                            <li
                                {...renderOptionProps.getOptionRecommendedProps()}
                            >
                                <Text>{option.label}</Text>
                                {isSelected && (
                                    <svg
                                        width='1rem'
                                        height='1rem'
                                        xmlns='http://www.w3.org/2000/svg'
                                        fill='none'
                                        viewBox='0 0 24 24'
                                        strokeWidth={1.5}
                                        stroke='currentColor'
                                        className='w-6 h-6'
                                    >
                                        <path
                                            strokeLinecap='round'
                                            strokeLinejoin='round'
                                            d='m4.5 12.75 6 6 9-13.5'
                                        />
                                    </svg>
                                )}
                            </li>
                        );
                    })}
                </ul>
            )}
        </div>
    );
};

export default Select;
