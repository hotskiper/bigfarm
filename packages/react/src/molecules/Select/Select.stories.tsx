import React from "react";
import Select from "./Select";

import "@ds.e/scss/lib/Select.css";

const options = [
    { label: "black", value: "black" },
    { label: "green", value: "green" },
    { label: "pink", value: "pink" },
];

export default {
    title: "Molecules/Select",
};

export const Common = () => <Select options={options} />;

export const RenderOption = () => (
    <Select
        options={options}
        renderOption={({ getOptionRecommendedProps, option, isSelected }) => (
            <span {...getOptionRecommendedProps()}>{option.label} {isSelected? 'selected' : ''}</span>
        )}
    />
);

export const CustomLabel = () => <Select label = 'select a color' options={options} />