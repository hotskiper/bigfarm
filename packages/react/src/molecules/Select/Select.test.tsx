import React from "react";
import Select from "./Select";
import { render, fireEvent } from "@testing-library/react";

const options = [
    { label: "black", value: "black" },
    { label: "green", value: "green" },
    { label: "pink", value: "pink" },
];

test("renders all options passed to it", () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);
    fireEvent.click(getByTestId("DseSelectButton"));
    expect(getAllByRole("menuitemradio")).toHaveLength(options.length);
});

test("renders options using custom renderOption method if passed as prop", () => {
    const { getAllByTestId,getByTestId } = render(
        <Select
            options={options}
            renderOption={({ option }) => {
                return <li key={option.value} data-testid='CustomRenderOption'>{option.label}</li>;
            }}
        />
    );
    fireEvent.click(getByTestId("DseSelectButton"));
    expect(getAllByTestId('CustomRenderOption')).toHaveLength(options.length);
});

test("calls the onOptionSelected prop with the selected option and its index if passed", () => {
    const onOptionSelected = jest.fn();
    const { getAllByRole, getByTestId } = render(<Select options={options} onOptionSelected={onOptionSelected} />);
    fireEvent.click(getByTestId("DseSelectButton"));
    fireEvent.click(getAllByRole('menuitemradio')[0])
    //toHaveBeenCalledWith验证一个函数是否以预期的参数被调用
    expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0)
});

test("the button label changes to the selected option label", () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);
    fireEvent.click(getByTestId("DseSelectButton"));
    fireEvent.click(getAllByRole('menuitemradio')[0])
    expect(getByTestId("DseSelectButton")).toHaveTextContent(options[0].label);
});

test("snapshot of the slected option state", () => {
    const { getAllByRole, getByTestId, asFragment } = render(<Select options={options} />);
    fireEvent.click(getByTestId("DseSelectButton"));
    fireEvent.click(getAllByRole('menuitemradio')[0])
    expect(asFragment()).toMatchSnapshot();
});

test("snapshot of the base state", () => {
    const { asFragment } = render(<Select options={options} />);
    expect(asFragment()).toMatchSnapshot(); 
});

test("snapshot of the options menu open state", () => {
    const { getByTestId, asFragment } = render(<Select options={options} />);
    fireEvent.click(getByTestId("DseSelectButton"));
    expect(asFragment()).toMatchSnapshot();
});
