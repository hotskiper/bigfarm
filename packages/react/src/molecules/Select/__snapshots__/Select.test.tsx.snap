// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
      role="menu"
    >
      <div>
        Please select
      </div>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.5 15.75 7.5-7.5 7.5 7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
      role="menu"
    >
      <div>
        Please select
      </div>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.5 15.75 7.5-7.5 7.5 7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay"
      id="dse-select-list"
      style="top: 10px;"
    >
      <li
        aria-label="black"
        class="dse-select__option 
                                         
                                        
                                    "
        role="menuitemradio"
        tabindex="0"
      >
        <div>
          black
        </div>
      </li>
      <li
        aria-label="green"
        class="dse-select__option 
                                         
                                        
                                    "
        role="menuitemradio"
        tabindex="0"
      >
        <div>
          green
        </div>
      </li>
      <li
        aria-label="pink"
        class="dse-select__option 
                                         
                                        
                                    "
        role="menuitemradio"
        tabindex="0"
      >
        <div>
          pink
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the slected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
      role="menu"
    >
      <div>
        black
      </div>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.5 15.75 7.5-7.5 7.5 7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
